app/src/main/java/com/epfl/drawyourpath/userProfile/cache/UserModelCached.kt

```diff
@@ -0,0 +1,428 @@
+    fun setCurrentUser(userId: String) {
+        checkCurrentUser(false)
+        CompletableFuture.supplyAsync {
+            // insert empty user in cache to avoid null user
+            cache.insertIfEmpty(UserEntity(userId))
+        }.thenComposeAsync {
+            database.getUserAccount(userId)
+        }.thenApplyAsync {
+            cache.insert(fromUserModelToUserData(it))
+        }
+        setUserId(userId)
+    }
+
+    /**
+     * helper function for setting the userId
+     * @param userId the user id
+     */
+    private fun setUserId(userId: String) {
+        currentUserID = userId
+        _currentUserID.value = userId
+    }
+
+    /**
+     * get the current user (read-only).
+     *
+     * Its usage is as follow : `getUser().observe(viewLifecycleOwner) { viewText.text = it.username }`
+     * @return the [LiveData] of [UserEntity]
+     */
+    fun getUser(): LiveData<UserEntity> {
+        checkCurrentUser()
+        return user
+    }
+
+    /**
+     * get userModel
+     * @return the userModel
+     */
+    fun getUserModel(): UserModel {
+        return userModel
+    }
+
+    /**
+     * set the database used for online storing (mainly used for tests)
+     * @param database the database
+     */
+    fun setDatabase(database: Database) {
+        this.database = database
+        // if database is mock then set user to a test user
+        if (database is MockDataBase) {
+            setCurrentUser(database.userIdTest)
+        }
+    }
+
+    /**
+     * This function will return a future with a boolean to know if the username is available in the database
+     * (i.e the userName proposed is not already associated to another user profile)
+     * @param userName userName that the user want to use for his user profile
+     * @return the future that indicate if the username is available
+     */
+    fun isUsernameAvailable(userName: String): CompletableFuture<Boolean> {
+        return database.isUsernameAvailable(userName)
+    }
+
+    /**
+     * Use this function to modify the username(the username will be modify only if it is available on the database)
+     * @param username that we want to set
+     */
+    fun setUsername(username: String): CompletableFuture<Boolean> {
+        checkCurrentUser()
+        return CompletableFuture.supplyAsync {
+            database.setUsername(username).join()
+        }.thenApplyAsync {
+            if (it) {
+                cache.updateUsername(currentUserID!!, username)
+            }
+            it
+        }
+    }
+
+    /**
+     * Use this function to modify the daily distance goal of the user
+     * @param distanceGoal new daily distance goal
+     */
+    fun setDistanceGoal(distanceGoal: Double): CompletableFuture<Boolean> {
+        checkCurrentUser()
+        checkDistanceGoal(distanceGoal)
+        return CompletableFuture.supplyAsync {
+            database.setDistanceGoal(distanceGoal).join()
+        }.thenApplyAsync {
+            if (it) {

```

Comment:
I think it is incorrect , you should write instead : return database.setDistanceGoal(distanceGoal).thenApply{ isSet ->
if(isSet) cache.updateDistanceGoal(currentUserID!!, distanceGoal)
isSet
}